File: warrior_perception/insta360_camera/insta360_camera/insta360_node.py
"""USB / UVC driver node for the Insta360 X4 / X5.

When the camera is connected by USB-C with webcam mode enabled, it enumerates
as a standard UVC device (``/dev/videoN``). This node opens it with OpenCV,
grabs frames, and republishes them as ``sensor_msgs/Image``.
"""

import cv2
import rclpy
from cv_bridge import CvBridge
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import Image


class Insta360Node(Node):
    def __init__(self) -> None:
        super().__init__('insta360_node')

        self.declare_parameter('device', '/dev/video0')
        self.declare_parameter('frame_id', 'insta360_optical_frame')
        self.declare_parameter('width', 2880)
        self.declare_parameter('height', 1440)
        self.declare_parameter('fps', 30.0)

        self.device = self.get_parameter('device').get_parameter_value().string_value
        self.frame_id = self.get_parameter('frame_id').get_parameter_value().string_value
        self.width = self.get_parameter('width').get_parameter_value().integer_value
        self.height = self.get_parameter('height').get_parameter_value().integer_value
        self.fps = self.get_parameter('fps').get_parameter_value().double_value

        self.bridge = CvBridge()
        self.pub = self.create_publisher(Image, 'camera/image_raw', qos_profile_sensor_data)

        self.cap = cv2.VideoCapture(self.device, cv2.CAP_V4L2)
        if not self.cap.isOpened():
            self.get_logger().error(f'failed to open {self.device}')
            raise RuntimeError(f'cannot open {self.device}')
        self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, self.width)
        self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, self.height)
        self.cap.set(cv2.CAP_PROP_FPS, self.fps)

        period = 1.0 / max(self.fps, 1.0)
        self.timer = self.create_timer(period, self._tick)
        self.get_logger().info(
            f'insta360_node streaming {self.device} '
            f'@ {self.width}x{self.height}/{self.fps:.0f}fps'
        )

    def _tick(self) -> None:
        ok, frame = self.cap.read()
        if not ok:
            self.get_logger().warn('frame grab failed')
            return
        msg = self.bridge.cv2_to_imgmsg(frame, encoding='bgr8')
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = self.frame_id
        self.pub.publish(msg)

    def destroy_node(self) -> bool:
        if self.cap is not None:
            self.cap.release()
        return super().destroy_node()


def main() -> None:
    rclpy.init()
    node = Insta360Node()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: warrior_perception/insta360_camera/src/insta360_node.cpp
// ROS 2 C++ driver for the Insta360 X4 / X5 in USB/UVC webcam mode.
//
// Discovery: scans /sys/class/video4linux/ and matches by USB vendor ID
// (0x2e1a) or a case-insensitive "insta360" name substring — never assumes
// a fixed /dev/videoN path (port numbers change on replug, see CLAUDE.md Rule 1).
//
// Reconnects automatically after MAX_CONSECUTIVE_FAILURES missed frames.

#include <algorithm>
#include <cctype>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <string>

#include <opencv2/imgproc.hpp>
#include <opencv2/videoio.hpp>

#include <cv_bridge/cv_bridge.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>

namespace fs = std::filesystem;

static constexpr const char* INSTA360_VID         = "2e1a";
static constexpr int         MAX_CONSEC_FAILURES   = 15;

// ── helpers ───────────────────────────────────────────────────────────────────

static std::string read_sysfs(const fs::path& p)
{
    std::ifstream f(p);
    if (!f.is_open()) return {};
    return {std::istreambuf_iterator<char>(f), {}};
}

static std::string to_lower(std::string s)
{
    std::transform(s.begin(), s.end(), s.begin(), ::tolower);
    return s;
}

// Returns "/dev/videoN" for the first Insta360 UVC device found, or "".
static std::string find_insta360()
{
    const fs::path root = "/sys/class/video4linux";
    std::error_code ec;
    if (!fs::exists(root, ec)) return {};

    std::vector<fs::path> entries;
    for (auto& e : fs::directory_iterator(root, ec))
        entries.push_back(e.path());
    std::sort(entries.begin(), entries.end());

    for (const auto& dir : entries) {
        // Match by USB vendor ID in the uevent file
        if (to_lower(read_sysfs(dir / "device" / "uevent")).find(INSTA360_VID)
                != std::string::npos)
            return "/dev/" + dir.filename().string();

        // Fallback: v4l2 device name substring
        if (to_lower(read_sysfs(dir / "name")).find("insta360")
                != std::string::npos)
            return "/dev/" + dir.filename().string();
    }
    return {};
}

// ── node ─────────────────────────────────────────────────────────────────────

class Insta360Node : public rclcpp::Node
{
public:
    Insta360Node() : Node("insta360_node")
    {
        declare_parameter("device",   std::string(""));   // "" = auto-discover
        declare_parameter("frame_id", std::string("insta360_optical_frame"));
        declare_parameter("width",    2880);
        declare_parameter("height",   1440);
        declare_parameter("fps",      30.0);

        explicit_device_ = get_parameter("device").as_string();
        frame_id_        = get_parameter("frame_id").as_string();
        width_           = get_parameter("width").as_int();
        height_          = get_parameter("height").as_int();
        fps_             = get_parameter("fps").as_double();

        pub_ = create_publisher<sensor_msgs::msg::Image>(
            "camera/image_raw", rclcpp::SensorDataQoS());

        if (!open_device())
            RCLCPP_WARN(get_logger(), "Insta360 not found at startup — will retry");

        const double period_s = 1.0 / std::max(fps_, 1.0);
        timer_ = create_wall_timer(
            std::chrono::duration<double>(period_s),
            [this]() { tick(); });
    }

private:
    // ── device management ─────────────────────────────────────────────────────

    std::string resolve_device()
    {
        if (!explicit_device_.empty()) return explicit_device_;

        const std::string dev = find_insta360();
        if (dev.empty())
            RCLCPP_ERROR(get_logger(),
                "No Insta360 UVC device found — check webcam mode is enabled "
                "and the USB-C cable is seated");
        return dev;
    }

    bool open_device()
    {
        const std::string dev = resolve_device();
        if (dev.empty()) return false;

        auto cap = std::make_unique<cv::VideoCapture>(dev, cv::CAP_V4L2);
        if (!cap->isOpened()) {
            RCLCPP_ERROR(get_logger(), "VideoCapture failed to open %s", dev.c_str());
            return false;
        }

        // Force MJPEG before setting resolution/fps — YUYV (the default) is
        // limited to ~10 fps at 1920×1080 by USB bandwidth; MJPEG lifts that.
        cap->set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M','J','P','G'));
        cap->set(cv::CAP_PROP_FRAME_WIDTH,  static_cast<double>(width_));
        cap->set(cv::CAP_PROP_FRAME_HEIGHT, static_cast<double>(height_));
        cap->set(cv::CAP_PROP_FPS,          fps_);

        cap_      = std::move(cap);
        failures_ = 0;

        RCLCPP_INFO(get_logger(),
            "Streaming %s @ %dx%d / %.0f fps", dev.c_str(), width_, height_, fps_);
        return true;
    }

    void release_device()
    {
        cap_.reset();
    }

    // ── timer callback ────────────────────────────────────────────────────────

    void tick()
    {
        if (!cap_) {
            open_device();
            return;
        }

        cv::Mat frame;
        if (!cap_->read(frame) || frame.empty()) {
            if (++failures_ >= MAX_CONSEC_FAILURES) {
                RCLCPP_WARN(get_logger(),
                    "%d consecutive frame failures — releasing handle and re-discovering",
                    MAX_CONSEC_FAILURES);
                release_device();
            }
            return;
        }

        failures_ = 0;

        std_msgs::msg::Header hdr;
        hdr.stamp    = now();
        hdr.frame_id = frame_id_;

        pub_->publish(*cv_bridge::CvImage(hdr, "bgr8", frame).toImageMsg());
    }

    // ── members ───────────────────────────────────────────────────────────────

    std::string explicit_device_;
    std::string frame_id_;
    int         width_{1920};
    int         height_{1080};
    double      fps_{30.0};

    std::unique_ptr<cv::VideoCapture>                     cap_;
    int                                                   failures_{0};
    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr                          timer_;
};

// ── main ─────────────────────────────────────────────────────────────────────

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Insta360Node>());
    rclcpp::shutdown();
}
